pynbody/filt.py

```diff
@@ -100,36 +100,50 @@ class Sphere(Filter):
     def __init__(self, radius, cen=(0, 0, 0)):
         self._descriptor = "sphere"
         self.cen = np.asarray(cen)
-        if self.cen.shape != (3,):
-            raise ValueError, "Centre must be length 3 array"
+#        if self.cen.shape != (3,):
+#            raise ValueError, "Centre must be length 3 array"
 
-        if isinstance(radius, str):
-            radius = units.Unit(radius)
+        if not isinstance(radius, list) : 
+            radius = [radius]
+
+        if isinstance(radius[0], str):
+            radius = map(units.Unit,radius)
+
+        # check if many different centers were passed in
+            if isinstance(cen, list) or isinstance(cen, np.ndarray) : 
+                if len(cen) != len(radius) : 
+                    radius = radius * len(cen)
 
         self.radius = radius
 
+
     def __call__(self, sim):
         radius = self.radius
 
         with sim.immediate_mode:
             pos = sim['pos']
 
-        if units.is_unit_like(radius):
-            radius = float(radius.in_units(pos.units,
-                                           **pos.conversion_context()))
+        if units.is_unit_like(radius[0]):
+            radius = np.asarray([r.in_units(pos.units, **pos.conversion_context()) for r in radius], dtype='float')
+
+        else : 
+            radius = np.asarray(radius, dtype='float')
 
         cen = self.cen
         if units.has_units(cen):
             cen = cen.in_units(pos.units)
+
+        if len(cen.shape) == 1 : 
+            cen = np.expand_dims(cen,axis=0)
+
         return util._sphere_selection(np.asarray(pos),np.asarray(cen,dtype=pos.dtype),radius)
 
 
     def __repr__(self):
-        if units.is_unit(self.radius):
-
-            return "Sphere('%s', %s)" % (str(self.radius), repr(self.cen))
+        if units.is_unit(self.radius[0]):
+            return " & ".join(["Sphere('%s', %s)" % (str(radius), repr(cen)) for radius, cen in zip(self.radius, self.cen)])
         else:
-            return "Sphere(%.2e, %s)" % (self.radius, repr(self.cen))
+            return " & ".join(["Sphere(%.2e, %s)" % (radius, repr(cen)) for radius, cen in zip(self.radius, self.cen)])
```

Comment:
sure! I don't really care what it looks like, either way the string representation is a complete mess so maybe we could do something sensible like show the first couple and then have '...' at the end. The user can still access the radii and centers from the object itself should it be needed. 


On Feb 6, 2015, at 10:04 PM, Andrew Pontzen <notifications@github.com> wrote:

> In pynbody/filt.py:
> 
> >          else:
> > -            return "Sphere(%.2e, %s)" % (self.radius, repr(self.cen))
> > +            return " & ".join(["Sphere(%.2e, %s)" % (radius, repr(cen)) for radius, cen in zip(self.radius, self.cen)])
> Actually wouldn't it just be neatest to be honest that it's a single operation and have repr return Sphere(ndarray(...),ndarray(...)) rather than pretend it's a string of operations?
> 
> —
> Reply to this email directly or view it on GitHub.
>